validator/lib/jsonapi/validator/resource.rb

```diff
@@ -0,0 +1,123 @@
+require 'jsonapi/validator/document'
+require 'jsonapi/validator/relationship'
+
+module JSONAPI
+  module Validator
+    class Resource
```

Comment:
Sinatra-param has a Sinatra dep, which is the same problem.
The idea is that the basic validation can happen in the jsonapi document directly, so that it can *then* be safely transformed into an arbitrary object, rather than making a first transformation for convenience and then re-transforming it into something useful, which is wasteful. I'm not saying it's the best approach but it's currently the one that makes most sense to me.